### PR TITLE
Index

### PR DESCRIPTION
**Pull request by mayabrandi**
This PR adds:

- A new index collection to the database
- A new index category collection to the database
- CLI for loading index data from lims into the index collection
- CLI for loading index category data from lims into the index category collection
- Two front end views for visualization of index data

## Back End
Is the data correctly loaded into the trending database?

### vogue load index_categories

The command:
`vogue load index_categories -a`
Should load all reagent labels with index category from lims into the index_category colleciton. 

The command runs without error:
 
Skärmavbild 2020-02-18 kl  15 03 31

The data is correctly loaded into the database. This is checked by comparing the labels in lims to what you get when running the following aggergatio directly in mongocompass:

Skärmavbild 2020-02-18 kl  14 59 38

Skärmavbild 2020-02-18 kl  14 57 14

### vogue load indexes
The command:
`vogue load indexes -a`
Checking data for sample ACC5739A1 index: A07 - UDI0049, which has been rerun several times 

Skärmavbild 2020-02-26 kl  11 21 03

run: HKWM7DSXX

Skärmavbild 2020-02-26 kl  11 09 34

Skärmavbild 2020-02-26 kl  11 43 09

Skärmavbild 2020-02-26 kl  11 49 19

 Okej/MB 26 feb 2020
run: HKY3VDSXX

Skärmavbild 2020-02-26 kl  11 14 11

Skärmavbild 2020-02-26 kl  11 44 27

Index had 0 target reads  on this run, so the run is not showing up in the plot

 Okej/MB 26 feb 2020

Run: HKH23DSXX

Skärmavbild 2020-02-26 kl  11 19 24

Skärmavbild 2020-02-26 kl  11 45 57

Skärmavbild 2020-02-26 kl  11 48 38

 Okej/MB 26 feb 2020

Run: HKHF7DSXX

Skärmavbild 2020-02-26 kl  11 24 46

Skärmavbild 2020-02-26 kl  11 35 43

Skärmavbild 2020-02-26 kl  11 48 10

 Okej/MB 26 feb 2020

## Front End
Are the views self explanatory? yes / @moahaegglund 
Is the data is correctly displayed in the plot/plots? Yes /MB

## Code Review
**Review:**
- [x] Validated and tested by @mayabrandi 
- [x] code approved by @adrosenbaum 
- [X] "Merge and deploy" approved by @hassanfa
Thanks for filling in who performed the code review and the test!

This [version](https://semver.org/) is a:
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

### CODE REVIEW

File: vogue/server/templates/reagent_label.html [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    Highcharts.chart('indexes', {
+    chart: {
+        type: 'column'
+    },
+    title: {
+        text:'<a href="{{ url_for('server.reagent_labels', index_category_url=index_category.replace(' ', '_')) }}">{{index_category}}</a>: {{reagent_label}}'
+    },
+    subtitle: {
+        text: 'Normalized index performance = (index_total_reads/flowcell_total_reads)/(index_target_reads/flowcell_target_reads)'
```
Comment (adrosenbaum) [inline vogue/server/templates/reagent_label.html]:
```suggestion
        text: 'Normalized index performance = (index_total_reads/index_target_reads)/(flowcell_total_reads/flowcell_target_reads)'
```

File: vogue/server/utils/reagent_labels.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
+from statistics import mean
+
+
+def reagent_category_data(adapter, index_categroy: str) -> dict:
```
Comment (adrosenbaum) [inline vogue/server/utils/reagent_labels.py]:
```suggestion
def reagent_category_data(adapter, index_category: str) -> dict:
```

File: vogue/server/utils/reagent_labels.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        }, {
+        '$match': {
+            'index_category.category': index_categroy}
+        }] + _normalized_performance_pipe()
+
+    aggregate_result = list(adapter.reagent_label_aggregate(pipe))
+
+    average_normalized_peformance = []
+
+    for data in aggregate_result:
```
Comment (adrosenbaum) [inline vogue/server/utils/reagent_labels.py]: Is this possible to do in mongo aggregation pipe?

File: vogue/server/utils/reagent_labels.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+def _build_normalized_index_performance_data(data: dict) -> list:
+    """Filtering out flowcells with low performance."""
+
+    normalized_peformance=[]
+    ziped_data = zip(data['flowcell_performance'], 
+                     data['normalized_index_performance'], 
+                     data['flowcell_id'])
+
+    for fc_performance, norm_performance, fc_id in ziped_data:
+        if fc_performance > 0.1:
```
Comment (adrosenbaum) [inline vogue/server/utils/reagent_labels.py]: Sure about 0.1 ?

File: vogue/server/utils/reagent_labels.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    return normalized_peformance
+
+
+def _build_normalized_index_performance_data(data: dict) -> list:
+    """Filtering out flowcells with low performance."""
+
+    normalized_peformance=[]
+    ziped_data = zip(data['flowcell_performance'], 
+                     data['normalized_index_performance'], 
+                     data['flowcell_id'])
```
Comment (adrosenbaum) [inline vogue/server/utils/reagent_labels.py]: Will these always be of same length?

File: vogue/server/views.py [hunk 6]
```diff
@@ -43,6 +42,12 @@ def index():
         return redirect(url_for('server.microsalt_st_time', year=year))
     if request.form.get('page') == 'genotype_plate':
         return redirect(url_for('server.genotype_plate'))
+    if request.form.get('page') == 'reagent_labels':
+        index_category_url = request.form.get('index_category').replace(' ','_')
+        return redirect(url_for('server.reagent_labels', index_category_url=index_category_url))
+    if request.form.get('page') == 'reagent_label':
+        name, category = request.form.get('reagent_label').split(',')
+        return redirect(url_for('server.reagent_label', reagent_label=name.replace(' ', ''), index_category=category))
```
Comment (adrosenbaum) [inline vogue/server/views.py]:
```suggestion
        return redirect(url_for('server.reagent_label', reagent_label=name.replace(' ', '_'), index_category=category))
```

File: vogue/server/utils/reagent_labels.py (hunk 5 above)
Comment (mayabrandi) [reply vogue/server/utils/reagent_labels.py]: yes

File: vogue/adapter/plugin.py [hunk 7]
```diff
@@ -211,3 +213,27 @@ def get_category(self, app_tag):
...
                                                {"category": 1})
         return tag.get('category') if tag else None
+
+    def get_reagent_label_category(self, reagent_label):
+        """Function get category based on application tag from the application tag collection"""
+        category = self.app_tag_collection.find_one({'name': reagent_label},
+                                                    {"category": 1})
+        return category.get('category') if category else None
+
+    def reagent_label_aggregate(self, pipe : list):
```
[the remark this answers is not in the document]
Comment (mayabrandi) [reply vogue/adapter/plugin.py]: Yes. Its easier to follow the code when the pipes are defined in utils.py, and passed to the aggregation functions as arguments.

File: vogue/server/utils/reagent_labels.py (hunk 3 above)
Comment (mayabrandi) [reply vogue/server/utils/reagent_labels.py]: Good point.

**Pull request merged**